File: erikdyrelius-python3/day13.py
from aocbase import readInput
import re
import functools

inp = readInput()
ex = """0: 3
1: 2
4: 4
6: 4"""

p = re.compile(r"(\d+): (\d+)")

def parseTable(inp):
    fw = list()
    for line in inp.splitlines():
        m = p.match(line)
        if m:
            fw.append((int(m.group(1)), (int(m.group(2))-1)*2))
        else:
            print(">>>",line)
    fw.sort(key=lambda x:x[1]*1000+x[0])
    return fw

def calcSeverity(fw):
    sev = 0
    for d, r in fw:
        if d % r == 0:
            sev += d*((r//2)+1)
    return sev

def findPassage(fw):
    delay = 0
    while True:
        safe = True
        for d, r in fw:
            if (d + delay) % r == 0:
                safe = False
                break
        if safe:
            break
        delay += 1
    return delay

print("Puzzle 1, example {}".format(calcSeverity(parseTable(ex))))
print("Puzzle 1, solution {}".format(calcSeverity(parseTable(inp))))
print("Puzzle 2, example {}".format(findPassage(parseTable(ex))))
print("Puzzle 2, solution {}".format(findPassage(parseTable(inp))))

File: erikdyrelius-nim/day13.nim
import aocbase
import unittest
import strutils
import sets
import algorithm
import re

type
    FireWall = tuple[dst: int, prd: int]

let
    pattern = re"^(\d+):\s+(\d+)$"

proc parseTable(inp: string): seq[FireWall] =
    result = @[]
    for line in inp.splitlines:
        if line =~ pattern:
            result.add((matches[0].parseInt, (matches[1].parseInt - 1) * 2))
        else:
            echo "Error-> ", line
    result.sort do (x, y: FireWall) -> int:
        result = cmp(x.prd, y.prd)
        if result == 0:
            result = cmp(x.dst, y.dst)

proc calcSeverity(fws: seq[FireWall]): int =
    for fw in fws:
        if (fw.dst mod fw.prd) == 0:
            result.inc(fw.dst * ((fw.prd div 2) + 1))

proc findPassage(fws: seq[FireWall]): int =
    while true:
        var safe = true
        for fw in fws:
            if ((fw.dst + result) mod fw.prd) == 0:
                safe = false
                break
        if safe:
            break
        result.inc

let inpData = getInpData()
let ex = """0: 3
1: 2
4: 4
6: 4"""

suite "Puzzle 1":
    test "Example":
        check(calcSeverity(parseTable(ex))==24)
    echo "Result: ", calcSeverity(parseTable(inpData))

suite "Puzzle 2":
    test "Example":
        check(findPassage(parseTable(ex)) == 10)
    echo "Result: ", findPassage(parseTable(inpData))